agent/fw_drupal8.c

```diff
@@ -605,12 +605,128 @@ NR_PHP_WRAPPER(nr_drupal94_invoke_all_with_clean) {
+
+  nr_php_string_hash_key_t* hook_key = NULL;
+  zval* hook_val = NULL;
+  nr_php_string_hash_key_t* class_key = NULL;
+  zval* class_val = NULL;
+  nr_php_string_hash_key_t* method_key = NULL;
+  zval* module_val = NULL;
+  zend_ulong key_num = 0;
+
+  char* hook_str = NULL;
+  char* class_str = NULL;
+  char* method_str = NULL;
+  char* module_str = NULL;
+  char* hookpath = NULL;
+
+  hook_implementation_map = nr_php_get_zval_object_property(
+      module_handler, "hookImplementationsMap");
+  if (hook_implementation_map) {
+    if (nr_php_is_zval_valid_array(hook_implementation_map)) {
+      ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(hook_implementation_map), key_num,
+                                hook_key, hook_val) {
+        (void)key_num;
+        if ((NULL == hook_key) || (0 == nr_php_is_zval_valid_array(hook_val))) {
+          nrl_warning(NRL_FRAMEWORK,
+                      "hookImplementationsMap[hook = %s]: invalid value",
+                      NRSAFESTR(ZEND_STRING_VALUE(hook_key)));
+          NR_FREE_HOOK_MEM
+          return false;
+        }
+
+        ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(hook_val), key_num, class_key,
+                                  class_val) {
+          (void)key_num;
+          if ((NULL == class_key)
+              || (0 == nr_php_is_zval_valid_array(class_val))) {
+            nrl_warning(NRL_FRAMEWORK,
+                        "hookImplementationsMap[class = %s]: invalid value",
+                        NRSAFESTR(ZEND_STRING_VALUE(class_key)));
+            NR_FREE_HOOK_MEM
+            return false;
+          }
+
+          ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(class_val), key_num, method_key,
+                                    module_val) {
+            (void)key_num;
+
+            NR_FREE_HOOK_MEM
+
+            if ((NULL == method_key)
+                || (0 == nr_php_is_zval_valid_string(module_val))) {
+              nrl_warning(NRL_FRAMEWORK,
+                          "hookImplementationsMap[method = %s]: invalid value",
+                          NRSAFESTR(ZEND_STRING_VALUE(method_key)));
+              NR_FREE_HOOK_MEM
+              return false;
+            }
+
+            hook_str = nr_strdup(ZEND_STRING_VALUE(hook_key));
+            class_str = nr_strdup(ZEND_STRING_VALUE(class_key));
+            method_str = nr_strdup(ZEND_STRING_VALUE(method_key));
+            module_str = nr_strdup(Z_STRVAL_P(module_val));
+
+            if (0
+                == nr_stricmp(ZEND_STRING_VALUE(class_key),
+                              "Drupal\\Core\\Extension\\ProceduralCall")) {
+              hookpath = nr_formatf("%s", method_str);
+            } else {
+              hookpath = nr_formatf("%s::%s", class_str, method_str);
+            }
+
+            nr_php_wrap_user_function_drupal(hookpath, nr_strlen(hookpath),
+                                             module_str, nr_strlen(module_str),
+                                             hook_str, nr_strlen(hook_str));
```

Comment:
Can't the memory be freed here right after this function is called?  So that is doesn't go through the loop again and strdup/nr_format values again that haven't had a chance to be freed?  If it's freed here right after it's been allocated, the NR_FREE_HOOK_MEM calls don't need to happen everywhere.

And if [this comment](https://github.com/newrelic/newrelic-php-agent/pull/1030/files#r1972391782) is addressed, then NR_FREE_HOOK_MEM isn't needed, and the only thing that would be needed would be one call to nr_free(hookpath) after nr_php_wrap_user_function_drupal.

Reply:
> So that is doesn't go through the loop again and strdup/nr_format values again that haven't had a chance to be freed?

How would that happen when the [first](https://github.com/newrelic/newrelic-php-agent/pull/1030/files?diff=unified&w=0#diff-6c636f6e3733aae919f4c4cf66ec5226065346587de8cbaf3918f483f9d0f523R671) operation in the loop is freeing the memory?

Reply:
Ah, true, but the first operation doesn't need to free the memory.
It can be simplified so that only hookpath needs to be freed, and that can be freed directly after using it.

If strdups are removed as suggested [here](https://github.com/newrelic/newrelic-php-agent/pull/1030/files#r1972391782), then NR_FREE_HOOK_MEM isn't needed, and the only thing that would be needed would be one call to nr_free(hookpath) after nr_php_wrap_user_function_drupal.

Reply:
> See if [3ded06d](https://github.com/newrelic/newrelic-php-agent/commit/3ded06d732c7cd7b4ee87b62f4d57a3dbbc60ef8) addresses your concern

See https://github.com/newrelic/newrelic-php-agent/pull/1030#discussion_r1973926151